Error when non-existent file is encountered

AnthonyBroadCrawford opened the pull request:
# PR for issue #3037 
This PR is now development complete and ready for review. 

## Decisions 
When implementing this feature I took the following approach 

1. Error and exit compilation if a non-existent file is encountered 
2. Create a new flag --ignore-missing that will continue compilation rather than error out when a missing file is encountered

## Questions 
Given I'm still fairly new to the codebase & Solidity in general I have the following questions 

1. In the method where I discover a missing file *and* the user hasn't passed in the `--ignore-missing` option I leverage `exit(1)` as a means to escape.  Primarily, it was the easiest path vs having to pass a boolean up the stack to inspect and then return 1/false further up the stack. 

2.  Please jump on my copy.  I don't have a feel for the prose throughout the code base for error message/communications within the compiler.  Open to any/all feedback. 

## Sample commands 
Here's a simple way to test it when compiling this branch.  This is executed from the ***build*** directory with the attached simple contract located in the same ***build*** directory 

*NO ERROR: * `./solc/solc --bin --ignore-missing test.sol missing-file.sol`
*ERROR* `./solc/solc --bin test.sol missing-file.sol`

## Sample contract 
Again, my above commands assume this is located in the *build* director 

```
pragma solidity ^0.4.17;

contract Test {

}
```

axic committed:
Warn if using address overloads on contracts

chriseth committed:
Use STATICCALL for pure function calls if EVM version supports it and 0.5.0 is activated.

chriseth committed:
Test for using staticcall for view and pure functions.

chriseth committed:
Document STATICCALL usage in experimental 0.5.0.

chriseth committed:
Warn about using loose inline assembly features as experimental 0.5.0 feature.

chriseth committed:
Changelog entry.

chriseth committed:
Tests for warnings/errors for loose assembly.

chriseth committed:
Tests for multi-dimensional arrays.

chriseth committed:
Fix multi-dimensional arrays in the ABI.

chriseth committed:
Changelog entry.

chriseth committed:
Use 0.4.21 pragma for documentation that uses "emit".

chriseth committed:
Assert that address members are not present on contract types in 0.5.0.

chriseth committed:
Use new eth binary for trusty.

chriseth committed:
Increase ccache size.

chriseth committed:
Clarify that the experimental flag is not set for "experimental 0.5.0"

ekpyron committed:
Add gasleft to FunctionType::richIdentifier().

ekpyron committed:
Use double quotes for suggestion about hex literals and denominations.

aj07 committed:
Some words on Remix

chriseth committed:
Fix test.

Commit:
Error when missing file and cli flag to support suppression of error

chriseth committed:
Update changelog for release.

chriseth committed:
Set version to 0.4.22.

grzegorzszczecin committed:
Fix a typo.

chriseth commented on solc/CommandLineInterface.cpp:
```diff
@@ -416,8 +418,13 @@ void CommandLineInterface::readInputFilesAndConfigureRemappings()
 				auto infile = boost::filesystem::path(path);
 				if (!boost::filesystem::exists(infile))
 				{
-					cerr << "Skipping non-existent input file \"" << infile << "\"" << endl;
```
Just turn this into an error. It should not be an option that you can switch on or off.

AnthonyBroadCrawford replied:
@chriseth we want to silently ignore the missing file?  Is there a reason we wouldn't to simply inform the user we are skipping it (for education, debugging, general UX)?  Not taking stance here, simply seeking to better understand your POV.  Thanks for all the feedback so far.

chriseth replied:
No, just stop processing. I'm not sure we need an option to allow files to be missing.

chriseth commented on solc/CommandLineInterface.cpp:
```diff
@@ -484,23 +491,33 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
+//
+//HERE - Error out instead of skipping non-existent source files #3037
+//
 void CommandLineInterface::createFile(string const& _fileName, string const& _data)
```
This is the routine that writes output files, you should not need to touch this.

AnthonyBroadCrawford replied:
@chriseth Yeah, I plan on removing that comment as it was a placeholder for me to simply orient myself in the code.

chriseth committed:
Properly skip cleanup if only enlarging storage array.

AnthonyBroadCrawford commented:
@chriseth I took your feedback and only error when a file is missing and command line flag to suppress is not provided.  Updated change log and squashed down to a single commit.  Let me know if you see anything else that needs changed.

AnthonyBroadCrawford commented:
Looking into the failing tests.

Commit:
fixed failing test due to missing --ignore-missing flag

axic commented:
Please use rebase and not merges.

axic commented on solc/CommandLineInterface.cpp:
```diff
@@ -416,8 +418,10 @@ void CommandLineInterface::readInputFilesAndConfigureRemappings()
+					}
 				}
 
 				if (!boost::filesystem::is_regular_file(infile))
```
I think this should be considered part of this. It just does skipping if "file is invalid".

AnthonyBroadCrawford replied:
@axic Ah, I can update that section as well.  To be clear, if the file is missing OR invalid I should communicate to the user and error out.  That's what you are requesting.

axic replied:
@chriseth are you also ok with that?

chriseth replied:
yes

AnthonyBroadCrawford replied:
@axic @chriseth Alright, officially added as scope.

chriseth commented on Changelog.md:
```diff
@@ -27,6 +27,9 @@ Features:
  * Syntax Checker: Issue warning when using overloads of ``address`` on contract instances.
  * Type Checker: disallow combining hex numbers and unit denominations as experimental 0.5.0 feature.
+ * Improved messaging when error spans multiple lines of a sourcefile 
+ * Commandline interface:  Error when missing file detected. Support to suppress with missing error ``--ignore-missing`` flag.
```
Please move this to the 0.4.22 section.

AnthonyBroadCrawford replied:
@chriseth Yup, my apologies.  I'll have an updated and squashed (removing merge) tonight.

chriseth commented on Changelog.md:
```diff
@@ -27,6 +27,9 @@ Features:
  * Syntax Checker: Issue error if no visibility is specified on contract functions as experimental 0.5.0 feature.
  * Syntax Checker: Issue warning when using overloads of ``address`` on contract instances.
  * Type Checker: disallow combining hex numbers and unit denominations as experimental 0.5.0 feature.
+ * Improved messaging when error spans multiple lines of a sourcefile 
```
This should not be part of this PR.

chriseth commented on solc/CommandLineInterface.cpp:
```diff
@@ -416,8 +418,10 @@ void CommandLineInterface::readInputFilesAndConfigureRemappings()
-					continue;
+					if(! m_args.count(g_argIgnoreMissingFiles)){
+						cerr << "Non-existent input file " << infile << endl;
+						exit(1);
```
Do other parts of the code directly call exit? If not, rather return a bool.

AnthonyBroadCrawford replied:
@chriseth  I couldn't find other instances so I can return a bool.  It's gonna cause a little more work further up the stack which is not a big deal, simply expands the scope for number of functions this change touches.  I'll have it return a bool.

chriseth commented on solc/CommandLineInterface.cpp:
```diff
@@ -487,20 +491,26 @@ bool CommandLineInterface::parseLibraryOption(string const& _input)
 	namespace fs = boost::filesystem;
+
 	// create directory if not existent
-	fs::path p(m_args.at(g_argOutputDir).as<string>());
```
Please do not touch this function in this PR.

AnthonyBroadCrawford replied:
@chriseth My apologies, I didn't mean to have those make it into the PR.  I will remove any touch to this function.

chriseth reviewed (changes requested):
@axic what is your opinion about having the switch?

Commit:
Error when non-existent or irregular file found

chriseth commented:
@AnthonyBroadCrawford this is still not properly rebased. Please tell us if we should do it for you.
